model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/updater/liquibase/MapJpaLiquibaseUpdaterProvider.java

```diff
@@ -0,0 +1,190 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.jpa.updater.liquibase;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.lang.reflect.Method;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import liquibase.Contexts;
+import liquibase.LabelExpression;
+import liquibase.Liquibase;
+import liquibase.changelog.ChangeSet;
+import liquibase.changelog.RanChangeSet;
+import liquibase.exception.LiquibaseException;
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.reflections.Reflections;
+import org.keycloak.connections.jpa.updater.liquibase.conn.LiquibaseConnectionProvider;
+import org.keycloak.connections.jpa.updater.liquibase.ThreadLocalSessionContext;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider;
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider.Status;
+
+public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
+
+    private static final Logger logger = Logger.getLogger(MapJpaLiquibaseUpdaterProvider.class);
+
+    private static final Map<Class<?>, String> MODEL_TO_CHANGELOG_FILE = new HashMap<>();
+    static {
+        MODEL_TO_CHANGELOG_FILE.put(ClientModel.class, "META-INF/jpa-clients-changelog.xml");
+    }
+
+    private final KeycloakSession session;
+
+    public MapJpaLiquibaseUpdaterProvider(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public void update(Class modelType, Connection connection, String defaultSchema) {
+        update(modelType, connection, null, defaultSchema);
+    }
+
+    @Override
+    public void export(Class modelType, Connection connection, String defaultSchema, File file) {
+        update(modelType, connection, file, defaultSchema);
+    }
+
+    private void update(Class modelType, Connection connection, File file, String defaultSchema) {
+        logger.debug("Starting database update");
+
+        // Need ThreadLocal as liquibase doesn't seem to have API to inject custom objects into tasks
+        ThreadLocalSessionContext.setCurrentSession(session);
+
+        Writer exportWriter = null;
```

Comment:
This functionality is copied from original updater. There it is used for manual migration, where migration sql script is generated instead of direct migration of database. I assumed we'd want to also have this possibility with new store implementations so I didn't remove it. Although I'm not sure if the assumption is correct :) wdyt?